test/rubygems/test_gem_update_suggestion.rb

```diff
@@ -0,0 +1,137 @@
+
+    Gem.ui.stub :tty?, tty do
+      Gem.stub :rubygems_version, rubygems_version do
+        Gem.stub :latest_rubygems_version, latest_rubygems_version do
+          cmd.stub :ci?, ci do
+            yield
+          end
+        end
+      end
+    end
+  ensure
+    Gem.configuration[:prevent_update_suggestion] = original_config
+    ENV["RUBYGEMS_PREVENT_UPDATE_SUGGESTION"] = original_env
+    Gem.disable_system_update_message = original_disable
+  end
+
+  def test_update_suggestion
+    Gem.stub :rubygems_version, Gem::Version.new("1.2.3") do
+      Gem.stub :latest_rubygems_version, Gem::Version.new("2.0.0") do
+        assert_equal @cmd.update_suggestion, <<~SUGGESTION
+
+          A new release of RubyGems is available: 1.2.3 → 2.0.0!
+          Run `gem update --system 2.0.0` to update your installation.
+
+        SUGGESTION
+      end
+    end
+  end
+
+  def test_eglible_for_update
+    with_eglible_environment(cmd: @cmd) do
+      Time.stub :now, 123456789 do
+        assert @cmd.eglible_for_update?
+        assert_equal Gem.configuration.last_update_check, 123456789
+
+        # test last check is written to config file
+        assert File.read(Gem.configuration.state_file_name).match("123456789")
+      end
+    end
+  end
+
+  def test_eglible_for_update_prevent_config
+    with_eglible_environment(cmd: @cmd) do
+      begin
+        original_config, Gem.configuration[:prevent_update_suggestion] = Gem.configuration[:prevent_update_suggestion], true
+        refute @cmd.eglible_for_update?
+      ensure
+        Gem.configuration[:prevent_update_suggestion] = original_config
+      end
+    end
+  end
+
+  def test_eglible_for_update_prevent_env
+    with_eglible_environment(cmd: @cmd) do
+      begin
+        original_env, ENV["RUBYGEMS_PREVENT_UPDATE_SUGGESTION"] = ENV["RUBYGEMS_PREVENT_UPDATE_SUGGESTION"], "yes"
+        refute @cmd.eglible_for_update?
+      ensure
+        ENV["RUBYGEMS_PREVENT_UPDATE_SUGGESTION"] = original_env
+      end
+    end
+  end
+
+  def test_eglible_for_update_non_tty
+    with_eglible_environment(tty: false, cmd: @cmd) do
+      refute @cmd.eglible_for_update?
+    end
+  end
+
+  def test_eglible_for_update_for_prerelease
+    with_eglible_environment(rubygems_version: Gem::Version.new("1.0.0-rc1"), cmd: @cmd) do
+      refute @cmd.eglible_for_update?
+    end
+  end
+
+  def test_eglible_for_update_disabled_update
+    with_eglible_environment(cmd: @cmd) do
+      begin
+        original_disable, Gem.disable_system_update_message = Gem.disable_system_update_message, "disabled"
+        refute @cmd.eglible_for_update?
+      ensure
+        Gem.disable_system_update_message = original_disable
+      end
+    end
+  end
+
+  def test_eglible_for_update_on_ci
+    with_eglible_environment(ci: true, cmd: @cmd) do
+      refute @cmd.eglible_for_update?
+    end
+  end
+
+  def test_eglible_for_update_unwrittable_config
+    with_eglible_environment(ci: true, cmd: @cmd) do
```

Comment:
Here we're not actually testing the code to detect unwritability, because we will early exit when checking the ci case due to `ci: true` above, correct? I guess that should be removed and probably some extra assertion about `state_file_writable?` actually being called?

Reply:
well spotted, fixed

Reply:
Nice. I was also suggesting somehow switching to mocks, not just stubs. If I'm understanding correct, if we have a mechanism that patches methods, like `Gem.configuration.stub :state_file_writable?, false`, but also makes sure the methods are actually called, then this test issue would not have been introduced in the first place because the original version was not calling `Gem.configuration.state_file_writable?` at all.

In any case, totally out of scope of this PR!

Reply:
I'm not sure about the historical reasons for rolling out custom simple stub library, but since we control test dependencies using Gemfile now, maybe we can sneak in some mock library. I can explore this if welcomed. Indeed out of this PRs scope.

Reply:
I think it was because of wanting to get rid of MiniTest and TestUnit not providing exact replacements, so we rolled our own. Feel free to explore using a mock library, sure!